### ISSUE TITLE
I2I: inline autocomplete

### ISSUE THREAD

**Issue by caroqliu** [ampproject/amphtml#24582]
## Summary
The [amp-autocomplete](https://amp.dev/documentation/components/amp-autocomplete/) extension replaces the entire input field with a selected value. AMP should provide an easy way for publishers to suggest and replace inline substrings that comprise a larger input value.

Example: +mention someone in an AMP Email

## Design document
[Link](https://docs.google.com/document/d/1iVJBQbNl2RTlshNqHSCEQ8zzXE8rg3NwSo3XKA1YwKs/edit?usp=sharing)

## Motivation
The existing component doesn't address use cases where the publisher wants to use an autocomplete-enabled input field to achieve more complex user interactions, such as suggesting a recommendation for part of a larger input value, or suggesting multiple possible values to form an aggregated selection value.

## Additional context
Different approaches are being weighed in the design doc.

/cc @ampproject/wg-ui-and-a11y @ampproject/wg-amp4email

**Comment by caroqliu** [comment]
Related #24522, [b/113159533](b/113159533)

**Comment by nainar** [comment]
Just putting in the requirements here: 

Requirements

- AMP for Websites
  - The ability to only replace part of the input field based on some kind of pattern matching, maybe a regex, as opposed to replacing the whole input field. 
- AMP for Ads
  - None currently
- AMP for Email
  - The ability to only replace part of the input field based on some kind of pattern matching, maybe a regex, as opposed to replacing the whole input field. 
    - E.g. Docs mentions use case, we need to only replace a token after the "+" sign.
  - A way to feed just part of the input field to the autocomplete endpoint, not the whole input field.
  - Need to intercept the XHRs to the autosuggest endpoint so that it can proxy it on the server side
- Stories AMP
  - None currently

MVP feature set: 

- inline - P0 - what does the MVP here look like?
  - Server side filter
  - Triggers for +/@ -> regular expressions
  - Styles the autocompleted result
    - Link
  - Keyboard interactions
    - Enter to set the value of the input field and trigger the Select action—This sets the value to the selected suggstion. 
    - Left (←) and Right (→) to navigate between characters in the input field.
    - Up (↑) and Down (↓) to navigate between suggested options. This means it changes the focus of the highlighted item, and if inline typeahead is enabled, the focused item is also displayed in the input field (see example below). These should loop.
    - Esc to hide filtered results box
  - token - P2

cc @caroqliu

**Comment by caroqliu** [comment]
Update: MVP feature set now excludes "styles the autocompleted result"

**Comment by dvoytenko** [comment]
@caroqliu sorry for the delay. Here's some feedback:

/1/ I mostly look at these inputs from the point of view of input and output type. For instance:
 - An autocomplete's input/output type is just a string, but it's probably something more of an enumerated shape. E.g. a city, or a product type.
 - A token autocomplete's input/output is an array of strings. From this point of view, it's more of an array of "autocompletes", though we'd definitely not want to present it this way. There's a precedent out there for a single component to support both a plain value and an array value. E.g. in React `<select multiple>` will use an array as a value type. But I think this is more of an exception than a rule.
 - An autocomplete-mention is a free form text entirely.

/2/ UX operations generally follow these types. E.g. a token autocomplete being an array would likely allow the wholesome deletion from the array via "x" in the UI. Etc.

Overall the /1/ and /2/ point to me that we do not get any obvious benefit from trying to either share the features in a fewer set of components or from composing them. I think these are three independent components. Any savings would come from reusing necessary code internally and from format consistency.

Another point of feedback is for mention-autocomplete. I believe we need to support match/value autocomplete response right away. This comes from plain-text type of the output in this component. For instance, the resulting text "+Harry Potter says expelliarmus would totally work on Voldemort" will not be usable. We have to replace "+Harry Potter" with "+hpotter@acme.com" to make mention parseable.

**Comment by caroqliu** [comment]
I agree with you that the token autocomplete case is a distinct one in terms of logic, data shape, and UX bells and whistles, so I'm going to respond only in terms of the single-input autocomplete and its inline counterpart. 

Regarding separating those two features into distinct components -- While there is meaningful difference at a high level regarding their return types, at the end of the day they both take and return plaintext strings, sometimes modifying the values. The user flow of seeing suggestions, navigating between them, and selecting from among them does not differ. 

From a devX perspective, isn't it better to keep functionality in one component when reasonably possible? @alanorozco and I recently discussed refactoring the implementation to bind divergent features to separate execution contexts, somewhat like what you did with `viewport-impl.js` to handle the differing needs of platforms. These could be initialized based on the presence of the `inline` attribute and makes maintenance over time of a consistent feature set more doable. Would something like that seem like a reasonable compromise to you?

To your comment about mention-autocomplete--I'm not sure I understand. If you mean that a user should be able to type "+Harry Potter" and select "+harrypotter@etc.com", that is possible now using rich-text templating. I will make a demo for this shortly.

**Comment by dvoytenko** [comment]
I think devX issues are easy to bridge if we ensure that the right code is being reused between the component. I think ultimately, the composition here on the dev level is slightly different than component level (i.e. the type->search->display-popup-with-values->navigate->select is the shared piece), but that's easy to remedy by the correct code reuse. But I still see autocomplete and inline autocomplete as semantically different: i.e. operating on mostly different data types even if it's just a string in the end.

**Comment by caroqliu** [comment]
Discussed offline - Token has been de-prioritized and there's not active work on the feature.

For the inline versus existing autocomplete features, the resolution was to register their functionality under one `amp-autocomplete`. The important acknowledgement here is that the input and output datatypes of the two modes have nuanced differences. Namely, the existing autocomplete has the opportunity to return one of an enumerated list of values, whereas the inline use case has the opportunity to return multiple from among an enumerated list of values. In practice, both types of input fields are treated as free form text with the option but not requirement of being modified by user selection from among an enumerated list of values. _However_, in supporting the new feature it's important to explicitly acknowledge that even if this were not the case, the difference in the way the two modes structure their data are not distinct _enough_ to warrant a separation at the component level.

Thanks @dvoytenko and @choumx for helping us more clearly articulate this subtle design decision.
